src/random-walk.js

```diff
@@ -41,64 +40,46 @@ class RandomWalk {
    */
   start () {
     // Don't run twice
-    if (this._running || !this._options.enabled) { return }
-
-    // Create running handle
-    const runningHandle = {
-      _onCancel: null,
-      _timeoutId: null,
-      runPeriodically: (walk, period) => {
-        runningHandle._timeoutId = setTimeout(() => {
-          runningHandle._timeoutId = null
-
-          walk((nextPeriod) => {
-            // Was walk cancelled while fn was being called?
-            if (runningHandle._onCancel) {
-              return runningHandle._onCancel()
-            }
-            // Schedule next
-            runningHandle.runPeriodically(walk, nextPeriod)
-          })
-        }, period)
-      },
-      cancel: (cb) => {
-        // Not currently running, can callback immediately
-        if (runningHandle._timeoutId) {
-          clearTimeout(runningHandle._timeoutId)
-          return cb()
-        }
-        // Wait to finish and then call callback
-        runningHandle._onCancel = cb
-      }
-    }
+    if (this._timeoutId || !this._options.enabled) { return }
 
     // Start doing random walks after `this._options.delay`
-    runningHandle._timeoutId = setTimeout(() => {
+    this._timeoutId = setTimeout(() => {
       // Start runner immediately
-      runningHandle.runPeriodically((done) => {
+      this._runPeriodically((done) => {
         // Each subsequent walk should run on a `this._options.interval` interval
         this._walk(this._options.queriesPerPeriod, this._options.timeout, () => done(this._options.interval))
       }, 0)
     }, this._options.delay)
-
-    this._runningHandle = runningHandle
   }
 
   /**
-   * Stop the random-walk process.
-   * @param {function(Error)} callback
+   * Stop the random-walk process. Any active
+   * queries will be aborted.
    *
    * @returns {void}
    */
-  stop (callback) {
-    const runningHandle = this._runningHandle
+  stop () {
+    clearTimeout(this._timeoutId)
+    this._timeoutId = null
+    this._controller && this._controller.abort()
+  }
 
-    if (!runningHandle) {
-      return callback()
-    }
+  /**
+   * Run function `walk` on every `interval` ms
+   * @param {function(callback)} walk The function to execute on `interval`
+   * @param {number} interval The interval to run on in ms
+   *
+   * @private
+   */
+  _runPeriodically (walk, interval) {
+    this._timeoutId = setTimeout(() => {
+      this._timeoutId = null
```

Comment:
I think setting this._timeoutId to null here means that start() can be called twice?

Reply:
Yes, good catch. I updated the double start test to better validate the internals which exposed the double start was indeed happening. Nullifying the timeoutId is now only done on stop, since it only needs to happen there.